### PR TITLE
JAVACLIENT-184: Add OAuth2 authentication and refresh mechanism

### PR DESCRIPTION
**Pull request by kevinleturc**

### CODE REVIEW

File: nuxeo-java-client/src/main/java/org/nuxeo/client/spi/auth/OAuth2AuthInterceptor.java [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+                               .newBuilder()
+                               .addHeader(HttpHeaders.AUTHORIZATION, "Bearer " + token.getAccessToken())
+                               .build();
+        return chain.proceed(request);
+    }
+
+    public static OAuth2AuthInterceptor obtainAuthFromAuthenticationCode(String baseUrl, String clientId, String code) {
+        OAuth2Client oAuth2Client = new OAuth2Client(baseUrl, clientId);
+        OAuth2Token token = oAuth2Client.fetchAccessTokenFromAuthenticationCode(code);
+        return new OAuth2AuthInterceptor(token, oAuth2Client);
```
Comment (troger) [inline nuxeo-java-client/src/main/java/org/nuxeo/client/spi/auth/OAuth2AuthInterceptor.java]: I would have simply returned:
```java
return obtainAuthFromAuthenticationCode(baseUrl, clientId, null, code);
```

File: nuxeo-java-client/src/main/java/org/nuxeo/client/spi/auth/OAuth2AuthInterceptor.java [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+            }
+        }
+        Request request = chain.request()
+                               .newBuilder()
+                               .addHeader(HttpHeaders.AUTHORIZATION, "Bearer " + token.getAccessToken())
+                               .build();
+        return chain.proceed(request);
+    }
+
+    public static OAuth2AuthInterceptor obtainAuthFromAuthenticationCode(String baseUrl, String clientId, String code) {
```
Comment (troger) [inline nuxeo-java-client/src/main/java/org/nuxeo/client/spi/auth/OAuth2AuthInterceptor.java]: It's an `authorizationCode`.

**Comment by efge** [comment]
@troger I haven't read the PR yet but retrying the call is not possible if a streamed body is sent...

File: nuxeo-java-client/src/main/java/org/nuxeo/client/spi/auth/oauth2/OAuth2Client.java [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+        }
+        RequestBody formBody = builder.build();
+        return executeRequest(formBody);
+    }
+
+    public OAuth2Token fetchAccessTokenFromJWT(String jwtToken) {
+        // create form body
+        FormBody.Builder builder = new FormBody.Builder(UTF_8)
+                                                              .add("grant_type",
+                                                                      "urn:ietf:params:oauth:grant-type:jwt-bearer")
```
Comment (efge) [inline nuxeo-java-client/src/main/java/org/nuxeo/client/spi/auth/oauth2/OAuth2Client.java]: Could you use constants for builder names and fixed values please?

File: nuxeo-java-client/src/main/java/org/nuxeo/client/spi/auth/oauth2/OAuth2Client.java [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+                                                              .add("refresh_token", token.getRefreshToken());
+        if (clientSecret != null) {
+            builder.add("client_secret", clientSecret);
+        }
+        RequestBody formBody = builder.build();
+        return executeRequest(formBody);
+    }
+
+    protected OAuth2Token executeRequest(RequestBody body) {
+        Request request = new Request.Builder().url(baseUrl + "oauth2/token").post(body).build();
```
Comment (efge) [inline nuxeo-java-client/src/main/java/org/nuxeo/client/spi/auth/oauth2/OAuth2Client.java]: Please use a constant for `"oauth2/token"`

File: nuxeo-java-client/src/main/java/org/nuxeo/client/spi/auth/oauth2/OAuth2Client.java [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+        try (Response response = CLIENT.newCall(request).execute()) {
+            if (!response.isSuccessful()) {
+                throw new NuxeoClientRemoteException(response.code(), response.message(), response.body().string(),
+                        null);
+            }
+            return MAPPER.readValue(response.body().charStream(), OAuth2Token.class);
+        } catch (IOException e) {
+            throw new NuxeoClientException("Error during call on Nuxeo server", e);
```
Comment (efge) [inline nuxeo-java-client/src/main/java/org/nuxeo/client/spi/auth/oauth2/OAuth2Client.java]: `on` -> `to`

**Comment by kevinleturc** [comment]
Requires https://github.com/nuxeo/nuxeo/pull/2819

**Comment by efge** [comment]
👍

**Pull request merged**